### ISSUE TITLE
Automated Testing?

### ISSUE THREAD

**Issue by L1lith** [qouteall/ImmersivePortalsMod#904]
This isn't a problem for me so if it's too much work/too much of a pain I understand. You might achieve better mod stability if you could implement automated testing. Here's one idea about how this potentially could be achieved:

Have a set of test worlds, accompanied with an expected visual output (given an xyz coordinate and specific screen resolution), then have the test functionality automatically spin up the worlds and compare them to the expected output. This would be neat because you could test complex interactions, and be automatically notified when small visual things break.

Another good test might just be to measure the frame-rate to make sure nothings lagging out super hard. This would be pretty useful, considering you recently implemented this neat feature
```
A new preview version for mc 1.18. This version contains an automatic performance adjustment mechanism. The client will record the average FPS and free memory amount. When the client performance is not good,  the client will render fewer chunks through portals, and the server will load fewer chunks indirectly to the player. This feature may improve performance on old PCs but how effective it is not yet known. I need feedback about how the portal lags in this version.
By lag I mean 1. lag when you approach or walk away from a portal  2. lag when seeing the portal  3. lag after crossing the portal
```

With an automated testing suite, you could request users run the test suite on their hardware and then they could send you the results, allowing you to get very meaningful understandings of how the mod functions on different computers :)

PS: it blows my mind how you can code all this crazy stuff

**Comment by qouteall** [comment]
Yes it's a good idea. Sometimes when I fix a bug, new bugs appear.
